src/charm.py

```diff
@@ -155,178 +189,117 @@ def __init__(self, *args):
             )
-        except lightkube.ApiError as e:
-            if e.status.code == 403:
-                self.on_deployed_without_trust()
-                return
-        # [
-        #    NodeAddress(address='192.168.0.228', type='InternalIP'),
-        #    NodeAddress(address='example.com', type='Hostname')
-        # ]
-        # Remember that OpenStack, for example, will return an internal hostname, which is not
-        # accessible from the outside. Give preference to ExternalIP, then InternalIP first
-        # Separated, as we want to give preference to ExternalIP, InternalIP and then Hostname
-        for typ in ["ExternalIP", "InternalIP", "Hostname"]:
-            for a in node.status.addresses:
-                if a.type == typ:
-                    return a.address
-
-    def _node_port(self, port_type: str) -> int:
-        """Return node port."""
-        service_name = f"{self.app.name}-nodeport"
-        try:
-            service = lightkube.Client().get(
-                lightkube.resources.core_v1.Service, service_name, namespace=self._namespace
-            )
-        except lightkube.ApiError as e:
-            if e.status.code == 403:
-                self.on_deployed_without_trust()
-                return
+        except lightkube.core.exceptions.ApiError as e:
             if e.status.code == 404:
-                return -1
-        # svc.spec.ports
-        # [ServicePort(port=6432, appProtocol=None, name=None, nodePort=31438, protocol='TCP', targetPort=6432)]
-        # Keeping this distinction, so we have similarity with mysql-router-k8s
-        if port_type == "rw" or port_type == "ro":
-            port = self.config.listen_port
-        else:
-            raise ValueError(f"Invalid {port_type=}")
-        logger.debug(f"Looking for NodePort for {port_type} in {service.spec.ports}")
-        for svc_port in service.spec.ports:
-            if svc_port.port == port:
-                return svc_port.nodePort
-        raise Exception(f"NodePort not found for {port_type}")
-
-    def get_all_k8s_node_hostnames_and_ips(
-        self,
-    ) -> Tuple[list[str], list[str]]:
-        """Return all node hostnames and IPs registered in k8s."""
-        try:
-            node = lightkube.Client().get(
-                lightkube.resources.core_v1.Node,
-                name=self._node_name,
-                namespace=self._namespace,
-            )
-        except lightkube.ApiError as e:
-            if e.status.code == 403:
-                self.on_deployed_without_trust()
-                return
-        hostnames = []
-        ips = []
-        for a in node.status.addresses:
-            if a.type in ["ExternalIP", "InternalIP"]:
-                ips.append(a.address)
-            elif a.type == "Hostname":
-                hostnames.append(a.address)
-        return hostnames, ips
+                return None
+            raise
+
+        return service
+
+    @functools.cache  # noqa: B019
```

Comment:
I'd prefer to pass the model name and move the function outside the class than disable the rule.

Reply:
addressed in https://github.com/canonical/pgbouncer-k8s-operator/commit/46b037d599cbddcda6fdaed26488c21bff6dc6d6